test/test_tools_add_pre_post_processing_to_model.py

```diff
@@ -158,24 +185,189 @@ def test_pytorch_superresolution(self):
+            # tokenizer = transformers.AutoTokenizer.from_pretrained("xlm-roberta-base")
+            tokenizer_parameters = TokenizerParam(
+                vocab_or_file=os.path.join(
+                    test_data_dir, "../sentencepiece.bpe.model"),
+                tweaked_bos_id=0,
+            )
+            tokenizer_impl = SentencePieceTokenizer(
+                tokenizer_parameters, add_eos=True, add_bos=True)
+        elif tokenizer_type == "BertTokenizer":
+            ref_output = [
+                np.array([[2, 236, 118, 16, 1566, 875, 643, 3]]),
+                np.array([[0, 0, 0, 0, 0, 0, 0, 0]]),
+                np.array([[1, 1, 1, 1, 1, 1, 1, 1]]),
+            ]
+            # tokenizer = transformers.AutoTokenizer.from_pretrained("lordtt13/emo-mobilebert")
+            tokenizer_parameters = TokenizerParam(vocab_or_file=os.path.join(
+                test_data_dir, "../bert.vocab"), do_lower_case=True)
+            tokenizer_impl = BertTokenizer(tokenizer_parameters)
+        elif tokenizer_type in ["HfBertTokenizer", "HfBertTokenizer_with_decoder"]:
+            ref_output = (
+                [
+                    np.array([[2, 236, 118, 16, 1566, 875, 643, 3,
+                             236, 118, 978, 1566, 875, 643, 3]]),
+                    np.array([[0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1]]),
+                    np.array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]),
+                ]
+                if tokenizer_type == "HfBertTokenizer"
+                else [np.array(["[CLS]"])]
+            )
+            # tokenizer = transformers.AutoTokenizer.from_pretrained("lordtt13/emo-mobilebert")
+            tokenizer_parameters = TokenizerParam(vocab_or_file=os.path.join(test_data_dir, "../hfbert.vocab"),
+                                                  do_lower_case=True, is_sentence_pair=True)
+            input_text = ("This is a test sentence",
+                          "This is another test sentence")
+            tokenizer_impl = BertTokenizer(tokenizer_parameters)
+        else:
+            raise Exception("Unknown tokenizer")
+
+        pipeline = pre_post_processing.PrePostProcessor(inputs)
+        # ref_output = list(tokenizer(*input_text, return_tensors="np").values())
+
+        pipeline.add_pre_processing([tokenizer_impl])
+        if tokenizer_type == "HfBertTokenizer_with_decoder":
+            pipeline.add_post_processing([
+                (BertTokenizerQADecoder(tokenizer_parameters), [
+                 pre_post_processing.utils.IoMapEntry("BertTokenizer", producer_idx=0, consumer_idx=2)])
+            ])
+            input_model = onnx.load(os.path.join(
+                test_data_dir, "../bert_qa_decoder_base.onnx"))
+        else:
+            input_model = onnx.ModelProto()
+            input_model.opset_import.extend(
+                [onnx.helper.make_operatorsetid("", 16)])
+        new_model = pipeline.run(input_model)
+        onnx.save_model(new_model, output_model)
```

Comment:
It feels like this part could be a separate function that is called by the tokenizer specific tests if refactored a little, and all the tokenizer specific logic to create tokenizer_impl could be in the specific tests. i.e. having a big if/elif/elif for all the tokenizers in a shared function seems unnecessary. 

The pipeline create/run code could be a function that takes args for pre-processor and post-processor instances and model. Don't add post-processing if post-processor instance was None. If no model provided create the empty one.

That would also mean the HfBertTokenizer_with_decoder specific logic on lines 265 to 271 would be in test_hfBert_tokenizer rather than shared code.

Reply:
Some of logic in create a pipeline could be re-used by the four test functions.
I have moved the others into their own test function.